nncf/quantization/algorithms/post_training/algorithm.py

```diff
@@ -188,28 +211,207 @@ def apply(
+        transformation_layout.register(command)
+        return model_transformer.transform(transformation_layout)
+
+    def _add_outputs_before_if_node(
+        self, model_transformer: ModelTransformer, model: TModel, if_node: NNCFNode
+    ) -> TModel:
+        """
+        Inserts extra outputs on If node inputs.
+
+        :param model_transformer: ModelTransformer instance.
+        :param model: Model instance.
+        :param if_node: If node.
+        :return: Model with extra outputs before If node.
+        """
+        transformation_layout = TransformationLayout()
+        for command in self._backend_entity.create_output_insertion_commands(model, if_node):
+            transformation_layout.register(command)
+        return model_transformer.transform(transformation_layout)
+
+    def _dfs_quantize_models(
+        self,
+        parent_model: TModel,
+        parent_graph: NNCFGraph,
+        parent_dataset: Dataset,
+        parent_statistic_points: Optional[StatisticPointsContainer],
+        parent_model_cnt: int,
+    ) -> Tuple[TModel, int]:
+        """
+
+        :param parent_model:
+        :param parent_graph:
+        :param parent_dataset:
+        :param parent_statistic_points:
+        :param parent_model_cnt:
+        :return:
+        """
+        if self._has_if_op(parent_graph, self._backend_entity.if_node_metatype):
+            model_transformer = factory.ModelTransformerFactory.create(parent_model)
+
+            global_model_cnt = parent_model_cnt
+            for if_node in parent_graph.get_nodes_by_metatypes([self._backend_entity.if_node_metatype]):
+                for if_submodel_condition in (True, False):
+                    model_cnt = global_model_cnt + 1
+                    child_model = self._extract_if_submodel(model_transformer, if_node, if_submodel_condition)
+                    child_model_input_names = self._backend_entity.get_if_subgraph_input_names(
+                        parent_model, if_node, if_submodel_condition
+                    )
+                    if_cond_input_name = self._backend_entity.get_if_cond_input_name(parent_model, if_node)
+                    parent_model_with_additional_outputs = self._add_outputs_before_if_node(
+                        model_transformer, parent_model, if_node
+                    )
+
+                    child_dataset = self._make_dataset_for_child_model(
+                        factory.EngineFactory.create(parent_model_with_additional_outputs),
+                        parent_dataset,
+                        if_cond_input_name,
+                        child_model_input_names,
+                        if_submodel_condition,
+                        model_cnt,
+                    )
+
+                    child_quantized_model, model_cnt = self._dfs_quantize_models(
+                        child_model, NNCFGraphFactory.create(child_model), child_dataset, None, model_cnt
+                    )
```

Comment:
IMO, it would be better to separate the `if` subgraphs extraction and quantization process to avoid recursion. As the positive effects of this: 

- Removing double checking of the `if` subgraphs' existence, since it's not possible to have `if` in the `if` (or possible?);
- The `_dfs_quantize_models` method looks difficult now, but separated methods are more convenient for understanding;
- It's not necessary to keep not-applicable to the child models options like `parent_statistic_points`;

Reply:
I am not sure what you mean by 

> to separate the if subgraphs extraction and quantization process to avoid recursion

extraction of IF subgraph for now is just one function. Please, give a suggestion by code or more detailed explanation.

> Removing double checking of the if subgraphs' existence, since it's not possible to have if in the if (or possible?);

Yes, it is possible that if node exists inside of IF else or then bodies.

> It's not necessary to keep not-applicable to the child models options like parent_statistic_points;

It is a good suggestion. It can be avoided by first quantizing the parent model and then quantizing its subgraphs. I will update its logic.

Reply:
> Please, give a suggestion by code or more detailed explanation.

Please, separate the `_dfs_quantize_models` method into pieces to avoid using recursion.

> Yes, it is possible that if node exists inside of IF else or then bodies.

Any examples? How deep can it be? I'm not sure that there is no limit to the recursion deep in this case.

Reply:
Update the main method body